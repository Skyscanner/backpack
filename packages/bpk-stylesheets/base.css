/*
 * 
 * Backpack - Skyscanner's Design System
 * 
 * Copyright 2018 Skyscanner Ltd
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * 
 *//*! normalize.css v8.0.0 | MIT License | github.com/necolas/normalize.css */html{line-height:1.15;-webkit-text-size-adjust:100%}body{margin:0}h1{font-size:2em;margin:.67em 0}hr{-webkit-box-sizing:content-box;box-sizing:content-box;height:0;overflow:visible}pre{font-family:monospace,monospace;font-size:1em}a{background-color:transparent}abbr[title]{border-bottom:none;text-decoration:underline;-webkit-text-decoration:underline dotted;text-decoration:underline dotted}b,strong{font-weight:bolder}code,kbd,samp{font-family:monospace,monospace;font-size:1em}small{font-size:80%}sub,sup{font-size:75%;line-height:0;position:relative;vertical-align:baseline}sub{bottom:-.25em}sup{top:-.5em}img{border-style:none}button,input,optgroup,select,textarea{font-family:inherit;font-size:100%;line-height:1.15;margin:0}button,input{overflow:visible}button,select{text-transform:none}[type=button],[type=reset],[type=submit],button{-webkit-appearance:button}[type=button]::-moz-focus-inner,[type=reset]::-moz-focus-inner,[type=submit]::-moz-focus-inner,button::-moz-focus-inner{border-style:none;padding:0}[type=button]:-moz-focusring,[type=reset]:-moz-focusring,[type=submit]:-moz-focusring,button:-moz-focusring{outline:1px dotted ButtonText}fieldset{padding:.35em .75em .625em}legend{-webkit-box-sizing:border-box;box-sizing:border-box;color:inherit;display:table;max-width:100%;padding:0;white-space:normal}progress{vertical-align:baseline}textarea{overflow:auto}[type=checkbox],[type=radio]{-webkit-box-sizing:border-box;box-sizing:border-box;padding:0}[type=number]::-webkit-inner-spin-button,[type=number]::-webkit-outer-spin-button{height:auto}[type=search]{-webkit-appearance:textfield;outline-offset:-2px}[type=search]::-webkit-search-decoration{-webkit-appearance:none}::-webkit-file-upload-button{-webkit-appearance:button;font:inherit}details{display:block}summary{display:list-item}[hidden],template{display:none}@-webkit-keyframes bpk-keyframe-spin{to{-webkit-transform:rotate(1turn);transform:rotate(1turn)}}@keyframes bpk-keyframe-spin{to{-webkit-transform:rotate(1turn);transform:rotate(1turn)}}html{font-size:100%;-webkit-box-sizing:border-box;box-sizing:border-box}*,:after,:before{-webkit-box-sizing:inherit;box-sizing:inherit}body{color:#524c61;font-family:-apple-system,BlinkMacSystemFont,Roboto,Oxygen,Ubuntu,Cantarell,Fira Sans,Droid Sans,Helvetica Neue,Arial,sans-serif;font-size:1rem;line-height:1.5rem}body.scaffold-font-size{font-size:13px}body.enable-font-smoothing{-webkit-font-smoothing:antialiased}.hidden,.hide{display:none!important}.visually-hidden,.visuallyhidden{position:absolute;width:1px;height:1px;margin:-1px;padding:0;border:0;overflow:hidden;clip:rect(0 0 0 0)}.visually-hidden.focusable:active,.visually-hidden.focusable:focus,.visuallyhidden.focusable:active,.visuallyhidden.focusable:focus{position:static;width:auto;height:auto;margin:0;overflow:visible;clip:auto}.invisible{visibility:hidden}.clearfix:after,.clearfix:before{content:"";display:table}.clearfix:after{clear:both}