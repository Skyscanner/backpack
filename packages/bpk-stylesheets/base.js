/******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, { enumerable: true, get: getter });
/******/ 		}
/******/ 	};
/******/
/******/ 	// define __esModule on exports
/******/ 	__webpack_require__.r = function(exports) {
/******/ 		if(typeof Symbol !== 'undefined' && Symbol.toStringTag) {
/******/ 			Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });
/******/ 		}
/******/ 		Object.defineProperty(exports, '__esModule', { value: true });
/******/ 	};
/******/
/******/ 	// create a fake namespace object
/******/ 	// mode & 1: value is a module id, require it
/******/ 	// mode & 2: merge all properties of value into the ns
/******/ 	// mode & 4: return value when already ns object
/******/ 	// mode & 8|1: behave like require
/******/ 	__webpack_require__.t = function(value, mode) {
/******/ 		if(mode & 1) value = __webpack_require__(value);
/******/ 		if(mode & 8) return value;
/******/ 		if((mode & 4) && typeof value === 'object' && value && value.__esModule) return value;
/******/ 		var ns = Object.create(null);
/******/ 		__webpack_require__.r(ns);
/******/ 		Object.defineProperty(ns, 'default', { enumerable: true, value: value });
/******/ 		if(mode & 2 && typeof value != 'string') for(var key in value) __webpack_require__.d(ns, key, function(key) { return value[key]; }.bind(null, key));
/******/ 		return ns;
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = "./index.js");
/******/ })
/************************************************************************/
/******/ ({

/***/ "./index.js":
/*!******************!*\
  !*** ./index.js ***!
  \******************/
/*! no exports provided */
/***/ (function(module, __webpack_exports__, __webpack_require__) {

"use strict";
eval("__webpack_require__.r(__webpack_exports__);\n/* harmony import */ var _index_scss__WEBPACK_IMPORTED_MODULE_0__ = __webpack_require__(/*! ./index.scss */ \"./index.scss\");\n/* harmony import */ var _index_scss__WEBPACK_IMPORTED_MODULE_0___default = /*#__PURE__*/__webpack_require__.n(_index_scss__WEBPACK_IMPORTED_MODULE_0__);\n/*\n * Backpack - Skyscanner's Design System\n *\n * Copyright 2016 Skyscanner Ltd\n *\n * Licensed under the Apache License, Version 2.0 (the \"License\");\n * you may not use this file except in compliance with the License.\n * You may obtain a copy of the License at\n *\n *   http://www.apache.org/licenses/LICENSE-2.0\n *\n * Unless required by applicable law or agreed to in writing, software\n * distributed under the License is distributed on an \"AS IS\" BASIS,\n * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.\n * See the License for the specific language governing permissions and\n * limitations under the License.\n */\n\n\n(function () {\n  if (typeof document === 'undefined') {\n    return;\n  }\n\n  var classNames = []; // touch support\n\n  classNames.push('ontouchstart' in window || window.DocumentTouch && document instanceof DocumentTouch // eslint-disable-line no-undef\n  ? 'touch-support' : 'no-touch-support'); // add more feature tests here...\n\n  document.documentElement.className += \" \".concat(classNames.map(function (className) {\n    return \"bpk-\".concat(className);\n  }).join(' '));\n})();\n\n//# sourceURL=webpack:///./index.js?");

/***/ }),

/***/ "./index.scss":
/*!********************!*\
  !*** ./index.scss ***!
  \********************/
/*! no static exports found */
/***/ (function(module, exports, __webpack_require__) {

eval("// extracted by mini-css-extract-plugin\n\n//# sourceURL=webpack:///./index.scss?");

/***/ })

/******/ });