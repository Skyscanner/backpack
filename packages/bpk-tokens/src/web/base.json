{
  "imports": [
    "./base/animations.json",
    "./base/autosuggest.json",
    "./base/badges.json",
    "./base/borders.json",
    "./base/box-shadows.json",
    "./base/breakpoints.json",
    "./base/buttons.json",
    "./base/calendar.json",
    "./base/cards.json",
    "./base/colors.json",
    "./base/forms.json",
    "./base/grids.json",
    "./base/horizontal-nav.json",
    "./base/icons.json",
    "./base/modals.json",
    "./base/scrims.json",
    "./base/notifications.json",
    "./base/panels.json",
    "./base/ratings.json",
    "./base/spacing.json",
    "./base/states.json",
    "./base/typography.json",
    "./base/z-index.json"
  ]
}
