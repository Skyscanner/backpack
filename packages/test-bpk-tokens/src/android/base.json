{
  "imports": [
    "./base/animations.json",
    "./base/borders.json",
    "./base/buttons.json",
    "./base/carousel-indicator-dots.json",
    "./base/colors.json",
    "./base/elevation.json",
    "./base/spacing.json",
    "./base/typography.json"
  ]
}
