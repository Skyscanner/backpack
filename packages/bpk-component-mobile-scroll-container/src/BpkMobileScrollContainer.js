/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import PropTypes from 'prop-types';
import debounce from 'lodash.debounce';
import React, { Component, type Node } from 'react';

import { cssModules, isRTL } from '../../bpk-react-utils';

import STYLES from './BpkMobileScrollContainer.module.scss';

const getClassName = cssModules(STYLES);

const computeScrollBarAwareHeight = (
  scrollerEl: ?HTMLElement,
  innerEl: ?HTMLElement,
): ?string => {
  if (!(scrollerEl && innerEl)) {
    return null;
  }

  const scrollBarVisibile = scrollerEl.offsetHeight - innerEl.offsetHeight > 0;
  return scrollBarVisibile ? `${innerEl.offsetHeight / 16}rem` : 'auto';
};

const computeScrollIndicatorClassName = (
  scrollerEl: ?HTMLElement,
  leadingIndicatorClassName: ?string = null,
  trailingIndicatorClassName: ?string = null,
) => {
  if (!scrollerEl) {
    return null;
  }

  const classNames = [];
  const { offsetWidth, scrollLeft, scrollWidth } = scrollerEl;

  const rtl = isRTL();
  const scrollValue = rtl ? -Math.floor(scrollLeft) : Math.ceil(scrollLeft);
  const showLeadingIndicator = scrollValue > 0;
  const showTrailingIndicator = scrollValue < scrollWidth - offsetWidth;
  const showLeftIndicator = rtl ? showTrailingIndicator : showLeadingIndicator;
  const showRightIndicator = rtl ? showLeadingIndicator : showTrailingIndicator;

  if (showLeftIndicator) {
    classNames.push(
      getClassName('bpk-mobile-scroll-container--left-indicator'),
    );
  }
  if (showRightIndicator) {
    classNames.push(
      getClassName('bpk-mobile-scroll-container--right-indicator'),
    );
  }
  if (showLeadingIndicator && leadingIndicatorClassName) {
    classNames.push(leadingIndicatorClassName);
  }
  if (showTrailingIndicator && trailingIndicatorClassName) {
    classNames.push(trailingIndicatorClassName);
  }

  return classNames;
};

type Props = {
  children: Node,
  innerContainerTagName: string,
  className: ?string,
  leadingIndicatorClassName: ?string,
  trailingIndicatorClassName: ?string,
  scrollerRef: ?Function,
  style: ?Object,
  showScrollbar: ?boolean,
};

type State = {
  computedHeight: ?string,
  scrollIndicatorClassName: ?string,
};

class BpkMobileScrollContainer extends Component<Props, State> {
  innerEl: ?HTMLElement;

  scrollerEl: ?HTMLElement;

  static propTypes = {
    children: PropTypes.node.isRequired,
    scrollerRef: PropTypes.func,
    innerContainerTagName: PropTypes.string,
    className: PropTypes.string,
    leadingIndicatorClassName: PropTypes.string,
    trailingIndicatorClassName: PropTypes.string,
    style: PropTypes.object, // eslint-disable-line react/forbid-prop-types
    showScrollbar: PropTypes.bool,
  };

  static defaultProps = {
    scrollerRef: null,
    innerContainerTagName: 'div',
    className: null,
    leadingIndicatorClassName: null,
    trailingIndicatorClassName: null,
    style: null,
    showScrollbar: false,
  };

  constructor() {
    super();

    this.state = {
      computedHeight: 'auto',
      scrollIndicatorClassName: null,
    };
  }

  componentDidMount() {
    this.setScrollBarAwareHeight();
    this.setScrollIndicatorClassName();
    window.addEventListener('resize', this.onWindowResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.onWindowResize);
  }

  onWindowResize = debounce(() => {
    this.setScrollBarAwareHeight();
    this.setScrollIndicatorClassName();
  }, 100);

  setScrollIndicatorClassName = () => {
    const classNames = computeScrollIndicatorClassName(
      this.scrollerEl,
      this.props.leadingIndicatorClassName,
      this.props.trailingIndicatorClassName,
    );

    if (!classNames) {
      return;
    }

    this.setState(() => ({
      scrollIndicatorClassName: classNames.join(' '),
    }));
  };

  setScrollBarAwareHeight = () => {
    if (this.props.showScrollbar) {
      return;
    }
    const computedHeight = computeScrollBarAwareHeight(
      this.scrollerEl,
      this.innerEl,
    );

    if (!computedHeight) {
      return;
    }

    this.setState(() => ({ computedHeight }));
  };

  render() {
    const classNames = [getClassName('bpk-mobile-scroll-container')];
    const {
      children,
      className,
      innerContainerTagName,
      leadingIndicatorClassName,
      scrollerRef,
      showScrollbar,
      style,
      trailingIndicatorClassName,
      ...rest
    } = this.props;

    if (className) {
      classNames.push(className);
    }
    if (this.state.scrollIndicatorClassName) {
      classNames.push(this.state.scrollIndicatorClassName);
    }

    const scrollerClassNames = getClassName(
      'bpk-mobile-scroll-container__scroller',
      showScrollbar && 'bpk-mobile-scroll-container__showScrollbar',
    );

    const InnerContainer = innerContainerTagName;

    return (
      <div
        {...rest}
        className={classNames.join(' ')}
        style={{ ...style, height: this.state.computedHeight }}
      >
        <div
          ref={(el) => {
            if (scrollerRef) {
              scrollerRef(el);
            }
            this.scrollerEl = el;
          }}
          onScroll={this.setScrollIndicatorClassName}
          className={scrollerClassNames}
        >
          <InnerContainer
            ref={(el) => {
              this.innerEl = el;
            }}
            className={getClassName('bpk-mobile-scroll-container__inner')}
          >
            {children}
          </InnerContainer>
        </div>
      </div>
    );
  }
}

export default BpkMobileScrollContainer;
export { computeScrollBarAwareHeight, computeScrollIndicatorClassName };
