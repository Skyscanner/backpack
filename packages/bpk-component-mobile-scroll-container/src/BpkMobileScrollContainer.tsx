/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { ReactNode, ElementType, CSSProperties } from 'react';
import { Component } from 'react';

import debounce from 'lodash/debounce';

import { cssModules, isRTL } from '../../bpk-react-utils';

import type { DebouncedFunc } from 'lodash';

import STYLES from './BpkMobileScrollContainer.module.scss';

const getClassName = cssModules(STYLES);

const computeScrollBarAwareHeight = (
  scrollerEl: HTMLElement | null,
  innerEl: HTMLElement | null,
) => {
  if (!(scrollerEl && innerEl)) {
    return null;
  }

  const scrollBarVisible = scrollerEl.offsetHeight - innerEl.offsetHeight > 0;
  return scrollBarVisible ? `${innerEl.offsetHeight / 16}rem` : 'auto';
};

const computeScrollIndicatorClassName = (
  scrollerEl: HTMLElement | null,
  leadingIndicatorClassName?: string,
  trailingIndicatorClassName?: string,
) => {
  if (!scrollerEl) {
    return null;
  }

  const classNames: string[] = [];
  const { offsetWidth, scrollLeft, scrollWidth } = scrollerEl;

  const rtl = isRTL();
  const scrollValue = rtl ? -Math.floor(scrollLeft) : Math.ceil(scrollLeft);
  const showLeadingIndicator = scrollValue > 0;
  const showTrailingIndicator = scrollValue < scrollWidth - offsetWidth;

  if (showLeadingIndicator && leadingIndicatorClassName) {
    classNames.push(leadingIndicatorClassName);
  }
  if (showTrailingIndicator && trailingIndicatorClassName) {
    classNames.push(trailingIndicatorClassName);
  }

  return classNames;
};

type Props = {
  ariaLabel?: string;
  children: ReactNode;
  innerContainerTagName?: ElementType;
  className?: string;
  leadingIndicatorClassName?: string;
  trailingIndicatorClassName?: string;
  scrollerRef?: (el: HTMLElement | null) => void;
  style?: CSSProperties;
  showScrollbar?: boolean;
};

type State = {
  computedHeight: string;
  scrollIndicatorClassName: string | null;
};

class BpkMobileScrollContainer extends Component<Props, State> {
  debouncedResize:  DebouncedFunc<() => void>

  static defaultProps: Partial<Props> = {
    innerContainerTagName: 'div',
    showScrollbar: false,
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      computedHeight: 'auto',
      scrollIndicatorClassName: null,
    };

    this.debouncedResize = debounce(this.onWindowResize, 100);
  }

  componentDidMount() {
    requestAnimationFrame(() => {
      this.setScrollBarAwareHeight();
      this.setScrollIndicatorClassName();
    });
    window.addEventListener('resize', this.debouncedResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.debouncedResize);
  }

  onWindowResize = () => {
    this.setScrollBarAwareHeight();
    this.setScrollIndicatorClassName();
  };

  setScrollIndicatorClassName = () => {
    const classNames = computeScrollIndicatorClassName(
      this.scrollerEl,
      this.props.leadingIndicatorClassName,
      this.props.trailingIndicatorClassName,
    );

    if (!classNames) {
      return;
    }

    this.setState(() => ({
      scrollIndicatorClassName: classNames.join(' '),
    }));
  };

  setScrollBarAwareHeight = () => {
    if (this.props.showScrollbar) {
      return;
    }
    const computedHeight = computeScrollBarAwareHeight(
      this.scrollerEl,
      this.innerEl,
    );

    if (!computedHeight) {
      return;
    }

    this.setState({ computedHeight });
  };

  innerEl: HTMLElement | null = null;

  scrollerEl: HTMLElement | null = null;

  render() {
    const classNames = [getClassName('bpk-mobile-scroll-container')];
    const {
      ariaLabel,
      children,
      className,
      innerContainerTagName = 'div',
      leadingIndicatorClassName,
      scrollerRef,
      showScrollbar,
      style,
      trailingIndicatorClassName,
      ...rest
    } = this.props;

    if (className) {
      classNames.push(className);
    }
    if (this.state.scrollIndicatorClassName) {
      classNames.push(this.state.scrollIndicatorClassName);
    }

    const scrollerClassNames = getClassName(
      'bpk-mobile-scroll-container__scroller',
      showScrollbar && 'bpk-mobile-scroll-container__showScrollbar',
    );

    const InnerContainer = innerContainerTagName;

    return (
      <div
        {...rest}
        className={classNames.join(' ')}
        style={{ ...style, height: this.state.computedHeight }}
      >
        <div
          ref={(el) => {
            if (scrollerRef) {
              scrollerRef(el);
            }
            this.scrollerEl = el;
          }}
          onScroll={this.setScrollIndicatorClassName}
          className={scrollerClassNames}
        >
          <InnerContainer
            aria-label={ariaLabel}
            ref={(el: any) => {
              this.innerEl = el;
            }}
            className={getClassName('bpk-mobile-scroll-container__inner')}
          >
            {children}
          </InnerContainer>
        </div>
      </div>
    );
  }
}

export default BpkMobileScrollContainer;
export { computeScrollBarAwareHeight, computeScrollIndicatorClassName };
