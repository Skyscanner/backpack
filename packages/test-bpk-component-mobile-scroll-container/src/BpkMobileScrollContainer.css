@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-mobile-scroll-container{position:relative;overflow:hidden}.bpk-mobile-scroll-container--left-indicator::before{position:absolute;top:0;bottom:0;content:' ';display:block;z-index:1;width:2.25rem;background-image:linear-gradient(90deg, #fff, rgba(255,255,255,0));pointer-events:none;left:0}.bpk-mobile-scroll-container--right-indicator::after{position:absolute;top:0;bottom:0;content:' ';display:block;z-index:1;width:2.25rem;background-image:linear-gradient(270deg, #fff, rgba(255,255,255,0));pointer-events:none;right:0}.bpk-mobile-scroll-container__scroller{white-space:nowrap;overflow-x:auto;overflow-y:hidden;-webkit-overflow-scrolling:touch;-ms-overflow-style:-ms-autohiding-scrollbar}.bpk-mobile-scroll-container__scroller::-webkit-scrollbar{display:none}.bpk-mobile-scroll-container__inner{width:100%}
