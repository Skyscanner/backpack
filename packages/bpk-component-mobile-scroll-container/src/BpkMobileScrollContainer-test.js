/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2019 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import React from 'react';
import renderer from 'react-test-renderer';
import { colorRed500 } from 'bpk-tokens/tokens/base.es6';

import BpkMobileScrollContainer, {
  computeScrollBarAwareHeight,
  computeScrollIndicatorClassName,
} from './BpkMobileScrollContainer';

const makeMockScroller = (
  scrollLeft,
  scrollWidth,
  offsetWidth,
  offsetHeight = 0,
): HTMLElement => {
  const element = document.createElement('div');
  Object.defineProperty(element, 'scrollLeft', { value: scrollLeft });
  Object.defineProperty(element, 'scrollWidth', { value: scrollWidth });
  Object.defineProperty(element, 'offsetWidth', { value: offsetWidth });
  Object.defineProperty(element, 'offsetHeight', { value: offsetHeight });
  return element;
};

const makeMockInnerEl = (offsetHeight): HTMLElement => {
  const element = document.createElement('div');
  Object.defineProperty(element, 'offsetHeight', { value: offsetHeight });
  return element;
};

describe('BpkMobileScrollContainer', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(
        <BpkMobileScrollContainer>
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean
          commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus
          et magnis dis parturient montes, nascetur ridiculus mus.
        </BpkMobileScrollContainer>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a custom "className" attribute', () => {
    const tree = renderer
      .create(
        <BpkMobileScrollContainer className="my-custom-class">
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean
          commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus
          et magnis dis parturient montes, nascetur ridiculus mus.
        </BpkMobileScrollContainer>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a custom "style" attribute', () => {
    const tree = renderer
      .create(
        <BpkMobileScrollContainer style={{ backgroundColor: colorRed500 }}>
          Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean
          commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus
          et magnis dis parturient montes, nascetur ridiculus mus.
        </BpkMobileScrollContainer>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('functions', () => {
    describe('computeScrollIndicatorClassName', () => {
      it('returns null if `scrollerEl is null', () => {
        expect(computeScrollIndicatorClassName(null)).toBeNull();
      });

      it('should return left className and custom leadingIndicatorClassName if scrolling left is possible', () => {
        const scrollerEl = makeMockScroller(10, 200, 200);

        expect(
          computeScrollIndicatorClassName(
            scrollerEl,
            'custom-leading-class-name',
            'custom-trailing-class-name',
          ),
        ).toEqual([
          'bpk-mobile-scroll-container--left-indicator',
          'custom-leading-class-name',
        ]);
      });

      it('should return right className and custom trailingIndicatorClassName if scrolling right is possible', () => {
        const scrollerEl = makeMockScroller(0, 250, 200);

        expect(
          computeScrollIndicatorClassName(
            scrollerEl,
            'custom-leading-class-name',
            'custom-trailing-class-name',
          ),
        ).toEqual([
          'bpk-mobile-scroll-container--right-indicator',
          'custom-trailing-class-name',
        ]);
      });

      it('should return right and left className plus custom leadingIndicatorClassName and custom trailingIndicatorClassName if scrolling both right and left is possible', () => {
        const scrollerEl = makeMockScroller(10, 250, 200);

        const classNames = computeScrollIndicatorClassName(
          scrollerEl,
          'custom-leading-class-name',
          'custom-trailing-class-name',
        );

        expect(classNames).toContain(
          'bpk-mobile-scroll-container--left-indicator',
        );
        expect(classNames).toContain(
          'bpk-mobile-scroll-container--right-indicator',
        );
        expect(classNames).toContain('custom-leading-class-name');
        expect(classNames).toContain('custom-trailing-class-name');
      });
    });

    describe('computeScrollBarAwareHeight', () => {
      it('should return null if scrollerEl is null', () => {
        expect(
          computeScrollBarAwareHeight(null, makeMockInnerEl(50)),
        ).toBeNull();
      });

      it('should return null if innerEl is null', () => {
        expect(
          computeScrollBarAwareHeight(makeMockScroller(10, 200, 200), null),
        ).toBeNull();
      });

      it('should return scroll bar aware height when scroll bar is visible', () => {
        const scrollerEl = makeMockScroller(0, 200, 200, 300);
        const innerEl = makeMockInnerEl(288);

        expect(computeScrollBarAwareHeight(scrollerEl, innerEl)).toBe('18rem');
      });

      it('should return scroll bar aware height when scroll bar is not visible', () => {
        const scrollerEl = makeMockScroller(0, 200, 200, 280);
        const innerEl = makeMockInnerEl(288);

        expect(computeScrollBarAwareHeight(scrollerEl, innerEl)).toBe('auto');
      });
    });
  });
});
