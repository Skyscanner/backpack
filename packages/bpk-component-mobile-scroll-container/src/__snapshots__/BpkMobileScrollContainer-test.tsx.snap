// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkMobileScrollContainer should render correctly 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <div
        class="bpk-mobile-scroll-container__inner"
      >
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkMobileScrollContainer should render correctly when showScrollbar is set to true 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller bpk-mobile-scroll-container__showScrollbar"
    >
      <div
        class="bpk-mobile-scroll-container__inner"
      >
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkMobileScrollContainer should render correctly with a custom "className" attribute 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container my-custom-class"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <div
        class="bpk-mobile-scroll-container__inner"
      >
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkMobileScrollContainer should render correctly with a custom "style" attribute 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container"
    style="background-color: rgb(209, 67, 91); height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <div
        class="bpk-mobile-scroll-container__inner"
      >
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkMobileScrollContainer should render correctly with ariaLabel prop 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <div
        aria-label="my nav content"
        class="bpk-mobile-scroll-container__inner"
      >
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BpkMobileScrollContainer should render correctly with ariaLabel prop set to undefined 1`] = `
<DocumentFragment>
  <div
    class="bpk-mobile-scroll-container"
    style="height: auto;"
  >
    <div
      class="bpk-mobile-scroll-container__scroller"
    >
      <div
        class="bpk-mobile-scroll-container__inner"
      >
        Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus.
      </div>
    </div>
  </div>
</DocumentFragment>
`;
